distributed/semaphore.py

```diff
@@ -411,34 +426,51 @@ async def create_semaphore():
 
         return create_semaphore().__await__()
 
+    def sync(self, func, *args, asynchronous=None, callback_timeout=None, **kwargs):
+        callback_timeout = parse_timedelta(callback_timeout)
+        if (
+            asynchronous
+            or self.asynchronous
+            or getattr(thread_state, "asynchronous", False)
+        ):
+            future = func(*args, **kwargs)
+            if callback_timeout is not None:
+                future = asyncio.wait_for(future, callback_timeout)
+            return future
+        else:
+            return sync(
+                self.io_loop, func, *args, callback_timeout=callback_timeout, **kwargs
+            )
```

Comment:
If we might want to replicate this for other objects then maybe we should add this method to the worker?

It can probably also be much simpler on the worker because there are only two cases:

1. On the event loop (easy to check)
2.  Inside of the ThreadPoolExecutor

Although of course the current implementation also probably works ok

Reply:
The API of this class still feels a bit awkward, especially in the `__init__` with the Client/Worker duality. I needed to introduce this to allow for the construction of this object on client side. In these cases I would not have access to an actual worker and I could not really use this method then if it was part of the Worker.

If I was certain to be on a worker, I might even go for the `Worker.batched_stream` instead of using dedicated connections removing the need to deal with async comm on this layer.

However, unless we can find a nice way to construct this object safely on client side, I think it is easier to keep as is. While I don't think this would need to be an actual functioning instance on the client side, implementing this might just introduce more complexity

Reply:
> The API of this class still feels a bit awkward, especially in the __init__ with the Client/Worker duality. I needed to introduce this to allow for the construction of this object on client side. In these cases I would not have access to an actual worker and I could not really use this method then if it was part of the Worker.

I wonder if we can make the client and worker look similar enough that we can use them interchangably here.  Adding a `sync` method to the worker is a good step in that direction.

Then we might be able to do things like 

```python
class Semaphore:
    def __init__(self, ...):
        self.server = get_server()  # client or worker, whichever is around
```

> If I was certain to be on a worker, I might even go for the Worker.batched_stream instead of using dedicated connections removing the need to deal with async comm on this layer.

The scheduler also has a BatchedSend object, but it's called `self.scheduler_comm`.  Maybe we should have an alias so that these are the same across the Client and Worker

```python
class Worker:
    @property
    def scheduler_comm(self):
        return self.batched_stream
```

(or we could rename `batched_stream` to `scheduler_comm`  directly.